File: apps/hrms/app/api/leave.py
from datetime import date, datetime, timezone
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Request, status
from pydantic import BaseModel, Field, model_validator
from sqlalchemy.orm import Session

from app.core.db import get_db
from app.core.security import active_user, require_roles
from app.models.core import (
    AuditLog, Employee, EmploymentStatus, Holiday, LeaveRequest, LeaveStatus,
    LeaveType, Role,
)
from app.services.leave import (
    approver_for, available_days, carry_forward, count_leave_days, find_overlap,
    holidays_for,
)

router = APIRouter(prefix="/api/leave", tags=["leave"])

MAX_BACKDATE_DAYS = 30


def _ip(request: Request):
    return request.client.host if request.client else None


# ---------------------------------------------------------------------------
# Schemas
# ---------------------------------------------------------------------------
class ApplyIn(BaseModel):
    leave_type_code: str
    start_date: date
    end_date: date
    half_day: bool = False
    reason: Optional[str] = Field(None, max_length=1000)
    document_path: Optional[str] = None

    @model_validator(mode="after")
    def check_range(self):
        if self.end_date < self.start_date:
            raise ValueError("End date must be on or after the start date.")
        return self


class DecisionIn(BaseModel):
    note: Optional[str] = Field(None, max_length=1000)


class RequestOut(BaseModel):
    id: int
    employee_id: int
    employee_name: str
    leave_type: str
    start_date: date
    end_date: date
    half_day: bool
    days_count: float
    status: LeaveStatus
    reason: Optional[str] = None
    approver_id: Optional[int] = None
    decision_note: Optional[str] = None


class BalanceOut(BaseModel):
    leave_type_code: str
    leave_type: str
    entitled_days: float
    accrued_days: float
    carried_forward_days: float
    taken_days: float
    available_days: float
    is_paid: bool


class HolidayIn(BaseModel):
    holiday_date: date
    name: str = Field(..., max_length=120)
    is_confirmed: bool = False


def _out(r: LeaveRequest, emp: Employee, lt: LeaveType) -> RequestOut:
    return RequestOut(
        id=r.id, employee_id=r.employee_id, employee_name=emp.full_name,
        leave_type=lt.name, start_date=r.start_date, end_date=r.end_date,
        half_day=r.half_day, days_count=r.days_count, status=r.status,
        reason=r.reason, approver_id=r.approver_id, decision_note=r.decision_note,
    )


def _can_view(actor: Employee, target: LeaveRequest) -> bool:
    if actor.role in (Role.hr, Role.admin) or actor.id == target.employee_id:
        return True
    return actor.role == Role.manager and target.approver_id == actor.id


# ---------------------------------------------------------------------------
# Employee-facing
# ---------------------------------------------------------------------------
@router.get("/types")
def list_types(db: Session = Depends(get_db), emp=Depends(active_user)):
    return [{"code": t.code, "name": t.name, "is_paid": t.is_paid,
             "annual_entitlement_days": t.annual_entitlement_days,
             "requires_document": t.requires_document}
            for t in db.query(LeaveType).order_by(LeaveType.id).all()]


@router.get("/balances", response_model=List[BalanceOut])
def balances(employee_id: Optional[int] = None, db: Session = Depends(get_db),
             actor=Depends(active_user)):
    target = actor
    if employee_id and employee_id != actor.id:
        if actor.role not in (Role.hr, Role.admin, Role.manager):
            raise HTTPException(403, "You can only see your own balances.")
        target = db.query(Employee).filter(Employee.id == employee_id).first()
        if not target:
            raise HTTPException(404, "No employee with that ID.")
        if actor.role == Role.manager and target.manager_id != actor.id:
            raise HTTPException(403, "You can only see your own team's balances.")

    from app.services.leave import ensure_balance_row

    out = []
    for lt in db.query(LeaveType).order_by(LeaveType.id).all():
        avail = available_days(db, target, lt)
        row = ensure_balance_row(db, target.id, lt.id, date.today().year)
        out.append(BalanceOut(
            leave_type_code=lt.code, leave_type=lt.name,
            entitled_days=row.entitled_days, accrued_days=row.accrued_days,
            carried_forward_days=row.carried_forward_days, taken_days=row.taken_days,
            available_days=avail, is_paid=lt.is_paid))
    db.commit()
    return out


@router.post("/apply", response_model=RequestOut, status_code=201)
def apply(body: ApplyIn, request: Request, db: Session = Depends(get_db),
          emp: Employee = Depends(active_user)):
    if emp.status == EmploymentStatus.exited:
        raise HTTPException(403, "This account is closed. Contact HR.")

    lt = db.query(LeaveType).filter(LeaveType.code == body.leave_type_code.upper()).first()
    if not lt:
        raise HTTPException(404, "That leave type doesn't exist.")

    if (date.today() - body.start_date).days > MAX_BACKDATE_DAYS:
        raise HTTPException(422,
                            f"You can't apply more than {MAX_BACKDATE_DAYS} days after "
                            "the leave started. Ask HR to record this for you.")

    if lt.requires_document and not body.document_path:
        raise HTTPException(422, f"{lt.name} needs a supporting document.")

    clash = find_overlap(db, emp.id, body.start_date, body.end_date)
    if clash:
        raise HTTPException(409,
                            f"You already have leave from {clash.start_date} to "
                            f"{clash.end_date}. Cancel it first.")

    try:
        days = count_leave_days(db, body.start_date, body.end_date, body.half_day)
    except ValueError as e:
        raise HTTPException(422, str(e))

    if days <= 0:
        raise HTTPException(422,
                            "Those dates are all weekends or public holidays — "
                            "no leave needed.")

    avail = available_days(db, emp, lt)
    if lt.is_paid and days > avail:
        raise HTTPException(409,
                            f"You have {avail} days of {lt.name} available "
                            f"and asked for {days}.")

    approver = approver_for(db, emp)
    if not approver:
        raise HTTPException(409, "No approver is set up. Contact HR.")

    req = LeaveRequest(employee_id=emp.id, leave_type_id=lt.id,
                       start_date=body.start_date, end_date=body.end_date,
                       half_day=body.half_day, days_count=days, reason=body.reason,
                       status=LeaveStatus.pending, approver_id=approver.id,
                       document_path=body.document_path)
    db.add(req)
    db.commit()
    db.refresh(req)
    db.add(AuditLog(actor_id=emp.id, action="leave.applied", entity="leave_request",
                    entity_id=str(req.id),
                    detail=f"{lt.code} {body.start_date}..{body.end_date} ({days}d)",
                    ip_address=_ip(request)))
    db.commit()
    return _out(req, emp, lt)


@router.get("/mine", response_model=List[RequestOut])
def my_requests(db: Session = Depends(get_db), emp: Employee = Depends(active_user)):
    rows = (db.query(LeaveRequest)
              .filter(LeaveRequest.employee_id == emp.id)
              .order_by(LeaveRequest.start_date.desc()).all())
    types = {t.id: t for t in db.query(LeaveType).all()}
    return [_out(r, emp, types[r.leave_type_id]) for r in rows]


@router.post("/{request_id}/cancel", response_model=RequestOut)
def cancel(request_id: int, request: Request, db: Session = Depends(get_db),
           emp: Employee = Depends(active_user)):
    req = db.query(LeaveRequest).filter(LeaveRequest.id == request_id).first()
    if not req:
        raise HTTPException(404, "No leave request with that ID.")
    if req.employee_id != emp.id and emp.role not in (Role.hr, Role.admin):
        raise HTTPException(403, "You can only cancel your own leave.")
    if req.status in (LeaveStatus.cancelled, LeaveStatus.rejected):
        raise HTTPException(409, "This request is already closed.")
    if req.status == LeaveStatus.approved and req.end_date < date.today() \
            and emp.role not in (Role.hr, Role.admin):
        raise HTTPException(409, "This leave has already been taken. Ask HR to amend it.")

    req.status = LeaveStatus.cancelled
    req.decided_at = datetime.now(timezone.utc)
    db.add(AuditLog(actor_id=emp.id, action="leave.cancelled", entity="leave_request",
                    entity_id=str(req.id), ip_address=_ip(request)))
    db.commit()
    db.refresh(req)
    owner = db.query(Employee).filter(Employee.id == req.employee_id).first()
    lt = db.query(LeaveType).filter(LeaveType.id == req.leave_type_id).first()
    return _out(req, owner, lt)


# ---------------------------------------------------------------------------
# Approver-facing
# ---------------------------------------------------------------------------
@router.get("/pending", response_model=List[RequestOut])
def pending(db: Session = Depends(get_db), actor=Depends(active_user)):
    q = db.query(LeaveRequest).filter(LeaveRequest.status == LeaveStatus.pending)
    if actor.role not in (Role.hr, Role.admin):
        q = q.filter(LeaveRequest.approver_id == actor.id)
    rows = q.order_by(LeaveRequest.start_date).all()
    emps = {e.id: e for e in db.query(Employee).all()}
    types = {t.id: t for t in db.query(LeaveType).all()}
    return [_out(r, emps[r.employee_id], types[r.leave_type_id]) for r in rows]


def _decide(request_id: int, approve: bool, note: Optional[str],
            request: Request, db: Session, actor: Employee) -> RequestOut:
    req = db.query(LeaveRequest).filter(LeaveRequest.id == request_id).first()
    if not req:
        raise HTTPException(404, "No leave request with that ID.")
    if req.status != LeaveStatus.pending:
        raise HTTPException(409, f"This request is already {req.status.value}.")
    if req.employee_id == actor.id:
        raise HTTPException(403, "You can't approve your own leave.")
    if actor.role not in (Role.hr, Role.admin) and req.approver_id != actor.id:
        raise HTTPException(403, "This request isn't assigned to you.")

    owner = db.query(Employee).filter(Employee.id == req.employee_id).first()
    lt = db.query(LeaveType).filter(LeaveType.id == req.leave_type_id).first()

    if approve and lt.is_paid:
        # Re-check at decision time — balance may have moved since the request.
        avail = available_days(db, owner, lt)
        if req.days_count > avail:
            raise HTTPException(409,
                                f"{owner.full_name} now has only {avail} days of "
                                f"{lt.name} available. Reject or ask them to shorten it.")

    req.status = LeaveStatus.approved if approve else LeaveStatus.rejected
    req.approver_id = actor.id
    req.decided_at = datetime.now(timezone.utc)
    req.decision_note = note
    db.add(AuditLog(actor_id=actor.id,
                    action="leave.approved" if approve else "leave.rejected",
                    entity="leave_request", entity_id=str(req.id),
                    detail=f"employee={owner.employee_code} {req.days_count}d",
                    ip_address=_ip(request)))
    db.commit()
    db.refresh(req)
    return _out(req, owner, lt)


@router.post("/{request_id}/approve", response_model=RequestOut)
def approve(request_id: int, body: DecisionIn, request: Request,
            db: Session = Depends(get_db),
            actor=Depends(require_roles("manager", "hr", "admin"))):
    return _decide(request_id, True, body.note, request, db, actor)


@router.post("/{request_id}/reject", response_model=RequestOut)
def reject(request_id: int, body: DecisionIn, request: Request,
           db: Session = Depends(get_db),
           actor=Depends(require_roles("manager", "hr", "admin"))):
    if not body.note:
        raise HTTPException(422, "Give a reason so the employee knows why.")
    return _decide(request_id, False, body.note, request, db, actor)


@router.get("/{request_id}", response_model=RequestOut)
def get_one(request_id: int, db: Session = Depends(get_db), actor=Depends(active_user)):
    req = db.query(LeaveRequest).filter(LeaveRequest.id == request_id).first()
    if not req:
        raise HTTPException(404, "No leave request with that ID.")
    if not _can_view(actor, req):
        raise HTTPException(403, "You don't have access to this.")
    owner = db.query(Employee).filter(Employee.id == req.employee_id).first()
    lt = db.query(LeaveType).filter(LeaveType.id == req.leave_type_id).first()
    return _out(req, owner, lt)


# ---------------------------------------------------------------------------
# Calendar and holidays
# ---------------------------------------------------------------------------
@router.get("/calendar/team")
def team_calendar(start: date, end: date, db: Session = Depends(get_db),
                  actor=Depends(active_user)):
    """Who's out, and when. Managers see their reports; HR sees everyone."""
    q = (db.query(LeaveRequest)
           .filter(LeaveRequest.status == LeaveStatus.approved,
                   LeaveRequest.start_date <= end, LeaveRequest.end_date >= start))
    if actor.role == Role.manager:
        ids = [e.id for e in db.query(Employee).filter(Employee.manager_id == actor.id)]
        q = q.filter(LeaveRequest.employee_id.in_(ids or [-1]))
    elif actor.role not in (Role.hr, Role.admin):
        q = q.filter(LeaveRequest.employee_id == actor.id)

    emps = {e.id: e for e in db.query(Employee).all()}
    types = {t.id: t for t in db.query(LeaveType).all()}
    return {
        "holidays": [{"date": h.holiday_date, "name": h.name, "confirmed": h.is_confirmed}
                     for h in db.query(Holiday)
                                .filter(Holiday.holiday_date >= start,
                                        Holiday.holiday_date <= end).all()],
        "leave": [{"employee": emps[r.employee_id].full_name,
                   "employee_id": r.employee_id,
                   "type": types[r.leave_type_id].name,
                   "start": r.start_date, "end": r.end_date, "days": r.days_count}
                  for r in q.order_by(LeaveRequest.start_date).all()],
    }


@router.get("/holidays/{year}")
def list_holidays(year: int, db: Session = Depends(get_db), actor=Depends(active_user)):
    rows = (db.query(Holiday).filter(Holiday.year == year)
              .order_by(Holiday.holiday_date).all())
    return [{"date": h.holiday_date, "name": h.name, "confirmed": h.is_confirmed}
            for h in rows]


@router.post("/holidays", status_code=201)
def add_holiday(body: HolidayIn, request: Request, db: Session = Depends(get_db),
                actor=Depends(require_roles("hr", "admin"))):
    exists = (db.query(Holiday)
                .filter(Holiday.holiday_date == body.holiday_date,
                        Holiday.name == body.name).first())
    if exists:
        raise HTTPException(409, "That holiday is already on the calendar.")
    h = Holiday(holiday_date=body.holiday_date, name=body.name,
                year=body.holiday_date.year, is_confirmed=body.is_confirmed)
    db.add(h)
    db.add(AuditLog(actor_id=actor.id, action="holiday.added", entity="holiday",
                    entity_id=str(body.holiday_date), detail=body.name,
                    ip_address=_ip(request)))
    db.commit()
    return {"date": h.holiday_date, "name": h.name, "confirmed": h.is_confirmed}


@router.post("/year-end/carry-forward")
def run_carry_forward(from_year: int, request: Request, db: Session = Depends(get_db),
                      actor=Depends(require_roles("hr", "admin"))):
    """Run once at year end. Rolls unused annual leave up to the policy cap."""
    results = []
    types = db.query(LeaveType).filter(LeaveType.accrues.is_(True)).all()
    for emp in db.query(Employee).filter(Employee.is_active.is_(True)).all():
        for lt in types:
            rolled = carry_forward(db, emp, lt, from_year)
            if rolled:
                results.append({"employee_id": emp.id, "leave_type": lt.code,
                                "days_carried": rolled})
    db.add(AuditLog(actor_id=actor.id, action="leave.carry_forward", entity="leave_balance",
                    entity_id=str(from_year), detail=f"{len(results)} balances rolled",
                    ip_address=_ip(request)))
    db.commit()
    return {"from_year": from_year, "rolled": results}

File: apps/web/src/services/hr/leave.ts
/**
 * Leave engine: entitlement, accrual, balance, overlap, approval chain.
 *
 * Ported from the original HRMS `app/services/leave.py` and `app/api/leave.py`.
 * The rules that matter and are easy to lose in a port:
 *   - weekends and public holidays never consume balance;
 *   - a paid request cannot exceed available balance, checked again at approval
 *     time because the balance moves between request and decision;
 *   - nobody approves their own leave.
 */
import { prisma, type TxClient } from '@/lib/db';
import { Conflict, Forbidden, NotFound } from '@/lib/errors';
import { audit } from '@/lib/security/audit';
import type { Ctx } from '@/lib/security/rbac';
import { annualLeaveAccrued, dayKey, daysBetween, toDay, workingDays } from './rules';
import { getHrPolicy, type HrPolicy } from './settings';
import { isApprover, isHrAdmin } from './access';
import { EMPLOYEE_WITH_PERSON } from './publicSelect';

// Re-exported so the many existing call sites keep one import for HR access.
export { isApprover, isHrAdmin };

/** Statuses that consume balance and block an overlapping request. */
export const BLOCKING_STATUSES = ['PENDING', 'APPROVED'] as const;

/** Roles at or above this rank can be the fallback approver. 10 is company_admin. */
const ADMIN_ROLE_RANK = 10;

type Db = TxClient | typeof prisma;

// ── Actor helpers ──────────────────────────────────────────────────────────

/** The acting user's own employee record. Not every workspace user has one. */
export async function myEmployee(ctx: Ctx, db: Db = prisma) {
  return db.employeeProfile.findFirst({
    where: { tenantId: ctx.tenantId, deletedAt: null, membership: { salesUserId: ctx.actor.id } },
  });
}

export async function requireEmployee(ctx: Ctx, employeeId: string, db: Db = prisma) {
  const employee = await db.employeeProfile.findFirst({
    where: { tenantId: ctx.tenantId, id: employeeId, deletedAt: null },
  });
  if (!employee) throw NotFound('Employee');
  return employee;
}

// ── Calendar ───────────────────────────────────────────────────────────────

export async function holidaysFor(ctx: Ctx, start: Date, end: Date, db: Db = prisma) {
  const rows = await db.hrHoliday.findMany({
    where: { tenantId: ctx.tenantId, holidayDate: { gte: toDay(start), lte: toDay(end) } },
    orderBy: { holidayDate: 'asc' },
  });
  return rows;
}

export async function countLeaveDays(
  ctx: Ctx,
  start: Date,
  end: Date,
  halfDay: boolean,
  db: Db = prisma,
  known?: HrPolicy,
) {
  const [holidayRows, policy] = await Promise.all([holidaysFor(ctx, start, end, db), known ?? getHrPolicy(ctx)]);
  const holidays = holidayRows.map((holiday) => holiday.holidayDate);
  if (halfDay) {
    if (dayKey(start) !== dayKey(end)) throw new Error('A half day must start and end on the same date.');
    return workingDays(start, end, holidays, policy.weekendDays) ? 0.5 : 0;
  }
  return workingDays(start, end, holidays, policy.weekendDays);
}

// ── Balances ───────────────────────────────────────────────────────────────

export async function ensureBalanceRow(
  ctx: Ctx,
  employeeId: string,
  leaveTypeId: string,
  year: number,
  db: Db = prisma,
) {
  const leaveType = await db.hrLeaveType.findFirst({ where: { tenantId: ctx.tenantId, id: leaveTypeId } });
  if (!leaveType) throw NotFound('Leave type');
  return db.hrLeaveBalance.upsert({
    where: { tenantId_employeeId_leaveTypeId_year: { tenantId: ctx.tenantId, employeeId, leaveTypeId, year } },
    update: {},
    create: { tenantId: ctx.tenantId, employeeId, leaveTypeId, year, entitledDays: leaveType.annualAllowance },
  });
}

/**
 * Accruing types (annual) build up month by month. Non-accruing types (sick,
 * maternity) are available in full from day one — they are event-driven, not earned.
 */
export async function recomputeAccrual(
  ctx: Ctx,
  employee: { id: string; joinedOn: Date | null },
  leaveType: { id: string; accrues: boolean; annualAllowance: number },
  asOf: Date,
  db: Db = prisma,
  known?: HrPolicy,
) {
  const year = toDay(asOf).getUTCFullYear();
  await ensureBalanceRow(ctx, employee.id, leaveType.id, year, db);
  const policy = known ?? (await getHrPolicy(ctx));
  const accruedDays = leaveType.accrues
    ? employee.joinedOn
      ? annualLeaveAccrued(employee.joinedOn, asOf, policy)
      : 0
    : leaveType.annualAllowance;

  return db.hrLeaveBalance.update({
    where: {
      tenantId_employeeId_leaveTypeId_year: {
        tenantId: ctx.tenantId,
        employeeId: employee.id,
        leaveTypeId: leaveType.id,
        year,
      },
    },
    data: { accruedDays },
  });
}

/** Accrued + carried forward − consumed, recomputed from the requests themselves. */
export async function availableDays(
  ctx: Ctx,
  employee: { id: string; joinedOn: Date | null },
  leaveType: { id: string; accrues: boolean; annualAllowance: number },
  asOf: Date = new Date(),
  db: Db = prisma,
  known?: HrPolicy,
) {
  const year = toDay(asOf).getUTCFullYear();
  const balance = await recomputeAccrual(ctx, employee, leaveType, asOf, db, known);

  const consuming = await db.hrLeaveRequest.findMany({
    where: {
      tenantId: ctx.tenantId,
      employeeId: employee.id,
      leaveTypeId: leaveType.id,
      status: { in: [...BLOCKING_STATUSES] },
      startDate: { gte: new Date(Date.UTC(year, 0, 1)), lt: new Date(Date.UTC(year + 1, 0, 1)) },
    },
    select: { days: true },
  });
  const takenDays = consuming.reduce((total, row) => total + row.days, 0);

  const updated = await db.hrLeaveBalance.update({
    where: {
      tenantId_employeeId_leaveTypeId_year: {
        tenantId: ctx.tenantId,
        employeeId: employee.id,
        leaveTypeId: leaveType.id,
        year,
      },
    },
    data: { takenDays },
  });

  return {
    balance: updated,
    available: round2(balance.accruedDays + updated.carriedForwardDays - takenDays),
  };
}

/** Every leave type with the employee's live position against it. */
export async function balancesFor(ctx: Ctx, employeeId: string, asOf: Date = new Date()) {
  const [employee, types, policy] = await Promise.all([
    requireEmployee(ctx, employeeId),
    prisma.hrLeaveType.findMany({ where: { tenantId: ctx.tenantId, isActive: true }, orderBy: { name: 'asc' } }),
    getHrPolicy(ctx),
  ]);

  return Promise.all(
    types.map(async (leaveType) => {
      const { balance, available } = await availableDays(ctx, employee, leaveType, asOf, prisma, policy);
      return {
        leaveTypeId: leaveType.id,
        code: leaveType.code,
        name: leaveType.name,
        paid: leaveType.paid,
        entitledDays: balance.entitledDays,
        accruedDays: balance.accruedDays,
        carriedForwardDays: balance.carriedForwardDays,
        takenDays: balance.takenDays,
        availableDays: available,
      };
    }),
  );
}

// ── Approval chain ─────────────────────────────────────────────────────────

/**
 * Line manager first. If there is none, or they have left, the highest-ranking
 * active administrator picks it up — a request with no approver can never be
 * decided, so falling back is not optional.
 */
export async function approverFor(
  ctx: Ctx,
  employee: { id: string; managerMembershipId: string | null },
  db: Db = prisma,
) {
  if (employee.managerMembershipId) {
    const manager = await db.employeeProfile.findFirst({
      where: {
        tenantId: ctx.tenantId,
        membershipId: employee.managerMembershipId,
        deletedAt: null,
        employmentStatus: { notIn: ['EXITED'] },
        membership: { status: 'ACTIVE' },
      },
    });
    if (manager) return manager;
  }

  // Filtered and ordered in the database. Loading every active employee to pick
  // one administrator in JavaScript is fine at ten staff and a problem at a
  // thousand, and it is the same query either way.
  return db.employeeProfile.findFirst({
    where: {
      tenantId: ctx.tenantId,
      deletedAt: null,
      id: { not: employee.id },
      employmentStatus: { notIn: ['EXITED'] },
      membership: {
        status: 'ACTIVE',
        salesUser: { status: 'ACTIVE', deletedAt: null, role: { rank: { lte: ADMIN_ROLE_RANK } } },
      },
    },
    orderBy: { membership: { salesUser: { role: { rank: 'asc' } } } },
  });
}

export async function findOverlap(
  ctx: Ctx,
  employeeId: string,
  start: Date,
  end: Date,
  excludeId?: string,
  db: Db = prisma,
) {
  return db.hrLeaveRequest.findFirst({
    where: {
      tenantId: ctx.tenantId,
      employeeId,
      status: { in: [...BLOCKING_STATUSES] },
      startDate: { lte: toDay(end) },
      endDate: { gte: toDay(start) },
      ...(excludeId ? { id: { not: excludeId } } : {}),
    },
  });
}

// ── Workflows ──────────────────────────────────────────────────────────────

export interface ApplyLeaveInput {
  employeeId?: string;
  leaveTypeId: string;
  startDate: Date;
  endDate: Date;
  halfDay?: boolean;
  reason?: string;
  documentId?: string;
}

export async function applyForLeave(ctx: Ctx, input: ApplyLeaveInput) {
  // An employee applies for themselves; HR may record leave on anyone's behalf.
  const self = await myEmployee(ctx);
  const employeeId = input.employeeId ?? self?.id;
  if (!employeeId) throw NotFound('Employee');
  if (employeeId !== self?.id && !isHrAdmin(ctx)) throw Forbidden('You can only apply for your own leave.');

  const employee = await requireEmployee(ctx, employeeId);
  if (employee.employmentStatus === 'EXITED') throw Forbidden('This account is closed. Contact HR.');

  const leaveType = await prisma.hrLeaveType.findFirst({ where: { tenantId: ctx.tenantId, id: input.leaveTypeId } });
  if (!leaveType) throw NotFound('Leave type');

  const start = toDay(input.startDate);
  const end = toDay(input.endDate);
  if (end < start) throw new Error('End date must be on or after the start date.');

  const policy = await getHrPolicy(ctx);
  const backdated = daysBetween(start, new Date());
  if (backdated > policy.leaveMaxBackdateDays && !isHrAdmin(ctx)) {
    throw Conflict(
      `You cannot apply more than ${policy.leaveMaxBackdateDays} days after the leave started. Ask HR to record it for you.`,
    );
  }
  if (leaveType.requiresDocument && !input.documentId) {
    throw Conflict(`${leaveType.name} needs a supporting document.`);
  }

  const clash = await findOverlap(ctx, employee.id, start, end);
  if (clash)
    throw Conflict(
      `Leave already exists from ${dayKey(clash.startDate)} to ${dayKey(clash.endDate)}. Cancel it first.`,
    );

  const days = await countLeaveDays(ctx, start, end, input.halfDay ?? false, prisma, policy);
  if (days <= 0) throw Conflict('Those dates are all weekends or public holidays — no leave is needed.');

  if (leaveType.paid) {
    const { available } = await availableDays(ctx, employee, leaveType, start, prisma, policy);
    if (days > available)
      throw Conflict(`${available} days of ${leaveType.name} are available and ${days} were requested.`);
  }

  const approver = await approverFor(ctx, employee);
  if (!approver) throw Conflict('No approver is set up for this employee. Contact HR.');

  const request = await prisma.hrLeaveRequest.create({
    data: {
      tenantId: ctx.tenantId,
      employeeId: employee.id,
      leaveTypeId: leaveType.id,
      startDate: start,
      endDate: end,
      days,
      halfDay: input.halfDay ?? false,
      reason: input.reason ?? null,
      documentId: input.documentId ?? null,
      status: 'PENDING',
      approverId: approver.id,
    },
    include: { leaveType: true, employee: EMPLOYEE_WITH_PERSON },
  });

  await audit(ctx, {
    event: 'RECORD_CREATED',
    objectType: 'hr_leave_request',
    recordId: request.id,
    metadata: { action: 'leave.applied', leaveType: leaveType.code, days, from: dayKey(start), to: dayKey(end) },
  });
  return request;
}

export async function decideLeave(ctx: Ctx, requestId: string, approve: boolean, note?: string) {
  if (!approve && !note) throw Conflict('Give a reason so the employee knows why the request was rejected.');
  if (!isApprover(ctx)) throw Forbidden('Your role does not allow leave decisions.');

  const request = await prisma.hrLeaveRequest.findFirst({
    where: { tenantId: ctx.tenantId, id: requestId },
    include: { employee: true, leaveType: true },
  });
  if (!request) throw NotFound('Leave request');
  if (request.status !== 'PENDING') throw Conflict(`This request is already ${request.status.toLowerCase()}.`);

  const self = await myEmployee(ctx);
  if (self && request.employeeId === self.id) throw Forbidden('You cannot approve your own leave.');
  if (!isHrAdmin(ctx) && request.approverId !== self?.id) throw Forbidden('This request is not assigned to you.');

  // The balance moves between application and decision — re-check before approving.
  if (approve && request.leaveType.paid) {
    const { available } = await availableDays(ctx, request.employee, request.leaveType, request.startDate);
    if (request.days > available) {
      throw Conflict(
        `Only ${available} days of ${request.leaveType.name} remain. Reject the request or ask for it to be shortened.`,
      );
    }
  }

  const decided = await prisma.hrLeaveRequest.update({
    where: { tenantId: ctx.tenantId, id: request.id },
    data: {
      status: approve ? 'APPROVED' : 'REJECTED',
      approverId: self?.id ?? request.approverId,
      decidedAt: new Date(),
      decisionNote: note ?? null,
    },
    include: { leaveType: true, employee: EMPLOYEE_WITH_PERSON },
  });

  // Keep the stored balance in step with the decision straight away.
  await availableDays(ctx, request.employee, request.leaveType, request.startDate);

  await audit(ctx, {
    event: 'RECORD_UPDATED',
    objectType: 'hr_leave_request',
    recordId: decided.id,
    previousValue: { status: request.status },
    newValue: { status: decided.status },
    metadata: { action: approve ? 'leave.approved' : 'leave.rejected', days: decided.days },
  });
  return decided;
}

export async function cancelLeave(ctx: Ctx, requestId: string) {
  const request = await prisma.hrLeaveRequest.findFirst({
    where: { tenantId: ctx.tenantId, id: requestId },
    include: { employee: true, leaveType: true },
  });
  if (!request) throw NotFound('Leave request');

  const self = await myEmployee(ctx);
  const own = self && request.employeeId === self.id;
  if (!own && !isHrAdmin(ctx)) throw Forbidden('You can only cancel your own leave.');
  if (request.status === 'CANCELED' || request.status === 'REJECTED') throw Conflict('This request is already closed.');
  if (request.status === 'APPROVED' && toDay(request.endDate) < toDay(new Date()) && !isHrAdmin(ctx)) {
    throw Conflict('This leave has already been taken. Ask HR to amend it.');
  }

  const cancelled = await prisma.hrLeaveRequest.update({
    where: { tenantId: ctx.tenantId, id: request.id },
    data: { status: 'CANCELED', cancelledAt: new Date(), cancelledById: self?.id ?? null },
    include: { leaveType: true, employee: EMPLOYEE_WITH_PERSON },
  });

  // Cancelling releases the days it was holding.
  await availableDays(ctx, request.employee, request.leaveType, request.startDate);

  await audit(ctx, {
    event: 'RECORD_UPDATED',
    objectType: 'hr_leave_request',
    recordId: cancelled.id,
    previousValue: { status: request.status },
    newValue: { status: 'CANCELED' },
    metadata: { action: 'leave.cancelled' },
  });
  return cancelled;
}

/**
 * Year end: unused annual leave rolls into the next year up to the policy cap.
 * Only accruing types carry forward — sick leave does not bank.
 */
export async function runCarryForward(ctx: Ctx, fromYear: number) {
  if (!isHrAdmin(ctx)) throw Forbidden('Only an administrator can run year-end carry-forward.');

  const [types, employees] = await Promise.all([
    prisma.hrLeaveType.findMany({ where: { tenantId: ctx.tenantId, accrues: true, isActive: true } }),
    prisma.employeeProfile.findMany({
      where: { tenantId: ctx.tenantId, deletedAt: null, employmentStatus: { notIn: ['EXITED'] } },
    }),
  ]);

  const rolled: { employeeId: string; leaveType: string; days: number }[] = [];
  for (const employee of employees) {
    for (const leaveType of types) {
      const previous = await prisma.hrLeaveBalance.findFirst({
        where: { tenantId: ctx.tenantId, employeeId: employee.id, leaveTypeId: leaveType.id, year: fromYear },
      });
      if (!previous) continue;

      const unused = Math.max(previous.accruedDays + previous.carriedForwardDays - previous.takenDays, 0);
      const days = round2(Math.min(unused, leaveType.maxCarryForward));
      if (days <= 0) continue;

      await ensureBalanceRow(ctx, employee.id, leaveType.id, fromYear + 1);
      await prisma.hrLeaveBalance.update({
        where: {
          tenantId_employeeId_leaveTypeId_year: {
            tenantId: ctx.tenantId,
            employeeId: employee.id,
            leaveTypeId: leaveType.id,
            year: fromYear + 1,
          },
        },
        data: { carriedForwardDays: days },
      });
      rolled.push({ employeeId: employee.id, leaveType: leaveType.code, days });
    }
  }

  await audit(ctx, {
    event: 'RECORD_UPDATED',
    objectType: 'hr_leave_balance',
    metadata: { action: 'leave.carry_forward', fromYear, balancesRolled: rolled.length },
  });
  return { fromYear, rolled };
}

/** Who is out and when, plus the holidays in the window. */
export async function teamCalendar(ctx: Ctx, start: Date, end: Date) {
  const self = await myEmployee(ctx);
  const scope = isHrAdmin(ctx)
    ? {}
    : isApprover(ctx)
      ? { OR: [{ approverId: self?.id ?? '' }, { employeeId: self?.id ?? '' }] }
      : { employeeId: self?.id ?? '' };

  const [leave, holidays] = await Promise.all([
    prisma.hrLeaveRequest.findMany({
      where: {
        tenantId: ctx.tenantId,
        status: 'APPROVED',
        startDate: { lte: toDay(end) },
        endDate: { gte: toDay(start) },
        ...scope,
      },
      include: { leaveType: true, employee: EMPLOYEE_WITH_PERSON },
      orderBy: { startDate: 'asc' },
    }),
    holidaysFor(ctx, start, end),
  ]);
  return { leave, holidays };
}

const round2 = (value: number) => Math.round(value * 100) / 100;
